### PR TITLE
[ETCM-763] [ETCM-819] improve fast sync stability

### PR DESCRIPTION
**Pull request by jvdp**
NB. this is based off of #977 

This PR includes:
* improved (debug) logging in FastSync to help diagnose issues
* add (unique) names for every actor (for the benefit of logging)
* fix edge cases where `Terminated` message wasn't handled and FastSync was stopped
* fix bug where skeleton handler actor failure wouldn't be recovered from, causing fast sync to get stuck (`currentSkeleton` and `skeletonHandler` are now cleared.)
* fix issue(s) caused by stale state / actors hanging around after branch resolution (every child actor is now killed and a new `SyncingHandler` is instantiated.)

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala [hunk 1]
```diff
@@ -115,6 +115,12 @@ class FastSync(
       }
   }
 
+  private var actorCount = 0
+  private def countActor: Int = {
+    actorCount += 1
+    actorCount
+  }
```
Comment (AnastasiiaL) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala]: what about
```
private val actorCounter = new java.util.concurrent.atomic.AtomicInteger  
private def countActor: Int = actorCounter.incrementAndGet

```

File: src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala [hunk 2]
```diff
@@ -362,19 +374,32 @@ class FastSync(
       }
     }
 
-    private def waitingForBranchResolution: Receive = handleStatus orElse {
+    private def waitingForBranchResolution: Receive = handleStatus orElse handleRequestFailure orElse {
       case FastSyncBranchResolverActor.BranchResolvedSuccessful(firstCommonBlockNumber, newMasterPeer) =>
+        log.debug(
+          s"resolved branch with first common block number $firstCommonBlockNumber for new master peer $newMasterPeer"
```
Comment (AnastasiiaL) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala]: nitpick: let's start logs messages with a capital letter

File: src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala [hunk 3]
```diff
@@ -362,19 +374,32 @@ class FastSync(
...
-          nextBlockToFullyValidate = firstCommonBlockNumber + 1
+        log.debug("starting with fresh SyncingHandler")
+        val syncingHandler = new SyncingHandler(
+          syncState.copy(
+            bestBlockHeaderNumber = firstCommonBlockNumber,
+            nextBlockToFullyValidate = firstCommonBlockNumber + 1,
+            pivotBlockUpdateFailures = 0
+          )
         )
-        masterPeer = Some(newMasterPeer)
```
Comment (AnastasiiaL) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala:375]: masterPeer gets updated elsewhere now?

File: src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala [hunk 4]
```diff
@@ -560,17 +588,23 @@ class FastSync(
       syncState = syncState.updateNextBlockToValidate(header, K, X)
     }
 
-    private def handleRewind(header: BlockHeader, peer: Peer, N: Int, duration: FiniteDuration): Unit = {
+    private def handleRewind(
+        header: BlockHeader,
+        peer: Peer,
+        N: Int,
+        duration: FiniteDuration,
+        process: Boolean = true
```
Comment (AnastasiiaL) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala]: `process` parameter name is a bit confusing. What would best describe the purpose of it? Maybe continueSync or concludeSync?

File: src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSyncBranchResolverActor.scala [hunk 5]
```diff
@@ -47,7 +47,11 @@ class FastSyncBranchResolverActor(
 
   private def waitingForPeerWithHighestBlock: Receive = handlePeerListMessages orElse { case StartBranchResolver =>
     getPeerWithHighestBlock match {
-      case Some(PeerWithInfo(peer, _)) => requestRecentBlockHeaders(peer, blockchain.getBestBlockNumber())
+      case Some(peerWithInfo @ PeerWithInfo(peer, _)) =>
+        log.debug(
+          s"starting branch resolution now with peer = $peerWithInfo and block number ${blockchain.getBestBlockNumber()}"
```
Comment (AnastasiiaL) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSyncBranchResolverActor.scala]:
```suggestion
          "Starting branch resolution now with peer {} and block number {}", peerWithInfo, blockchain.getBestBlockNumber()
```

File: src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala (hunk 2 above)
Comment (jvdp) [reply src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala]: Why not

File: src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala (hunk 4 above)
Comment (jvdp) [reply src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala]: Yeah `continueSyncing` should cover it.

File: src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala (hunk 3 above)
Comment (jvdp) [reply src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala:375]: Hmm actually no; it is selected again after branch resolution. I'll change this to be closer to the (old) intended behavior. 👍

File: src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala [hunk 6]
```diff
@@ -340,12 +348,13 @@ class FastSync(
...
           log.info("Max number of allowed failures reached. Switching branch and master peer.")
-          handleRewind(header, masterPeer.get, fastSyncBlockValidationN, blacklistDuration)
+
+          blockHeadersQueue.dequeueAll(_ => true)
+
+          handleRewind(header, masterPeer.get, fastSyncBlockValidationN, blacklistDuration, continueSyncing = false)
 
           // Start branch resolution and wait for response from the FastSyncBranchResolver actor.
           context become waitingForBranchResolution
-          currentSkeletonState = None
```
Comment (AnastasiiaL) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala:347]: the skeleton state should be kept now?

Comment (jvdp) [reply src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala:347]: It will be reset by starting from a fresh `SyncingHandler`.

**Review by AnastasiiaL — APPROVED** [review]

**Comment by leo-bogastry** [comment]
So one thing I have experienced when synching with Astor, was the fastSync failing without a proper error.
Here https://github.com/input-output-hk/mantis/pull/1001/files#diff-4b491aec4f7ba3bc8091f81d49100e0ccc38a2f396bcca38434118637e6c0befL637 when it calls `validateBlocks`, that method just returns `Valid` or `Invalid`.
So what do you think of also adding a log there?
This was my modification
```
def validateBlocks(
      requestedHashes: Seq[ByteString],
      blockBodies: Seq[BlockBody],
      log: LoggingAdapter
  ): BlockBodyValidationResult = {
    var result: BlockBodyValidationResult = Valid
    (requestedHashes zip blockBodies)
      .map { case (hash, body) => (blockchain.getBlockHeaderByHash(hash), body) }
      .forall {
        case (Some(header), body) =>
          val validationResult: Either[StdBlockValidator.BlockError, BlockValid] =
            validators.blockValidator.validateHeaderAndBody(header, body)
          result = validationResult.fold(
            error => {
              log.error("Block body validation failed with error {}", error)
              Invalid
            },
            _ => Valid
          )
          validationResult.isRight
        case _ =>
          result = DbError
          false
      }
    result
  }
```

**Review by AurelienRichez — APPROVED** [review]

**Comment by jvdp** [comment]
@LeonorLunatech thanks for the suggestion, I added the log message and also refactored it a bit.

**Pull request merged**